packages/chopsticks/src/plugins/run-block/index.ts

```diff
@@ -82,3 +86,231 @@ export const cli = (y: yargs.Argv) => {
+const zHash = z.string().length(66).and(zHex)
+
+const schema = z.object({
+  includeRaw: z.boolean().optional(),
+  includeParsed: z.boolean().optional(),
+  includeBlockDetails: z.boolean().optional(),
+  head: zHash.optional(),
+  block: z.object({
+    header: z.any(),
+    extrinsics: z.array(zHex),
+  }),
+})
+
+type Params = z.infer<typeof schema>
+
+export interface RunBlockParams {
+  /**
+   * Include raw storage diff. Default to true
+   */
+  includeRaw: Params['includeRaw']
+  /**
+   * Include parsed storage diff in json format
+   */
+  includeParsed: Params['includeParsed']
+  /**
+   * Include block details such as parsed extrinsics in json format
+   */
+  includeBlockDetails: Params['includeBlockDetails']
+  /**
+   * Set chain head before running the extrinsics
+   */
+  head: Params['head']
+  /**
+   * Block to run
+   */
+  block: Params['block']
+}
+
+/**
+ * The phase of an execution.
+ * `number` means the phase is ApplyExtrinsic and the value is the extrinsic index.
+ */
+export type Phase = 'Initialization' | 'Finalization' | number // extrinsic index
+
+export interface RunBlockResponse {
+  /**
+   * The storage diff of each phase.
+   */
+  phases: {
+    /**
+     * The phase of the execution. See {@link Phase}.
+     */
+    phase: Phase
+    /**
+     * Parsed storage diff. Only available when `includeParsed` is true.
+     */
+    parsed?: Record<string, Record<string, any>>
+    /**
+     * Raw storage diff. Only available when `includeRaw` is true.
+     */
+    raw?: [HexString, HexString | null][]
+    /**
+     * Runtime logs.
+     */
+    logs?: string[]
+  }[]
+  /**
+   * Block details. Only available when `includeBlockDetails` is true.
+   */
+  blockDetails?: {
+    /**
+     * Block timestamp in ms
+     */
+    timestamp?: string
+    /**
+     * Parsed events in this block.
+     */
+    events?: { phase: Phase; section: string; method: string; args: any[] }[]
+    /**
+     * Parsed extrinsics in this block.
+     */
+    extrinsics: {
+      section: string
+      method: string
+      args: any[]
+      success: boolean
+    }[]
+  }
+}
+
+export const name = 'runBlock'
+
+/**
+ * Run a set of extrinsics on top of a block and get the storage diff
+ * and optionally the parsed storage diff and block details.
+ * NOTE: The extrinsics should include inherents or tranasctions may have unexpected results.
+ *
+ * This function is a dev rpc handler. Use `dev_runBlock` as the method name when calling it.
+ */
+export const rpc = async ({ chain }: Context, [params]: [RunBlockParams]): Promise<RunBlockResponse> => {
+  const { includeRaw, includeParsed, includeBlockDetails, head, block } = schema.parse(params)
+
+  const includeRawStorage = includeRaw ?? true
+
+  if (head) {
+    const headBlock = await chain.getBlock(head)
+    if (!headBlock) {
+      throw Error(`Invalid block hash ${head}`)
+    }
+    await chain.setHead(headBlock)
```

Comment:
this doesn't seem right, why changing chain head?

Reply:
yeah no need to do it